NanoAmmo for Vehicles

Cerrik opened the pull request:
First PR for TGMC, expecting to need to rewrite a majority of this, so please let me know how to adjust for future contributions.
## About The Pull Request

Created a new wall vendor called the NanoAmmo.  The NanoAmmo is attached to every variation of the tadpole and inside every variation of the APC.  It uses the standard vendor UI with the added functionality of a maximum capacity for each item.  A compact ammo box (the ones vended from RO) can be consumed by the NanoAmmo to quickly replenish the corresponding ammunition without going above the maximum.

Marines can also use their regular ammo boxes to dump the contents into the NanoAmmo.

![NanoAmmo](https://github.com/user-attachments/assets/8716dab7-ed2b-48b0-86ad-92504b40739c)
Loading the NanoAmmo with AR-12 ammo boxes.  Note that it does not load above 60.

![ammunition box load](https://github.com/user-attachments/assets/f75943ed-499d-4ad3-bf75-2bed70f3b3bd)
Loading the NanoAmmo with a rainbow ammunition.  It only loads what wouldn't exceed the capacity.
## Why It's Good For The Game

Tadpoles and APCs tend to get filled with tens or hundreds of ammo boxes during the round, which can degrade performance and make it an unnecessary hassle for players to find their basic ammos.  The NanoAmmo encourages marines to store their basic ammunitions in the NanoAmmo for easier access and reduced clutter.

As the NanoAmmo can only accept the basic, infinite ammos dispensed from normal marine vendors, it can't store important ammunition like SG ammo, so a Shrike can still disrupt logistics with a sneaky brain blast.

Since the ammo in the NanoAmmo is not infinite, this also does not invalidate RO's importance of delivering ammo to the marines, as the marines will still need the RO to restock the NanoAmmo when it runs dry.
## Changelog

:cl:
add: Added the NanoAmmo vendor to tadpoles and APC
/:cl:

Cerrik commented:
The current version allows TE cells, TE powerpacks, and TE powerbackpacks to be stored, up to a ridiculous amount of 60.  I left them in because I considered them to be in that "basic ammo" category, but I have reservations given the TE field recharger also exists and this would partially negate it.  I'll need some feedback on the TE category.

xSlumLord commented:
speedmerge

Cerrik commented:
Converted back to draft while I make some more key additions/changes.

Cerrik commented:
Final additions made.  Can now load with regular ammo boxes and the NanoAmmo takes 10 seconds for a drone to acid, roughly 7.5 for a melter, and 5 for a queen.  I also refactored the code to be easier to read.

QualityVan commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,224 @@
+					qdel(I)
+					return
+	else if(istype(I, /obj/item/shotgunbox))
+		if(I.type == /obj/item/shotgunbox/blank || I.type == /obj/item/shotgunbox/clf_heavyrifle)
```
Special casing, why not just handle it by the vendor not having a record that matches?

Cerrik replied:
Will address in the next commit

QualityVan commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,224 @@
+/obj/machinery/vending/nanoammo/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/storage/box/visual/magazine/compact))	// If the item is a compact mag box (the Req ammo boxes)
+		var/storage_capacity = I.storage_datum.storage_slots
+		if(length(I.contents) < storage_capacity) // If the mag box is not full
```
What's the reason to refuse a partial instead of just loading as many as are there?

Cerrik replied:
Can adjust this one, sure

QualityVan commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,224 @@
+			for(var/datum/vending_product/checked_record AS in product_records + hidden_records + coin_records)	// Loop through vendor records to find a match
+				var/obj/item/ammo_magazine/M = checked_record.product_path
+				if(B.ammo_type == M.default_ammo)	// Found a match
+					if(checked_record.amount + 8 <= max_capacity[M.default_ammo])	// If filling the box wouldn't exceed the cap
```
Bad for the same reason, magic number assumes how much the box has (or what its max is, here) instead of just checking.

Cerrik replied:
This one cannot be easily changed.  Shotgun ammo boxes (from RO) store 200 shells at once, but are taken in handfuls.  Vending machines don't dispense handfuls of shells, they dispense shotgun shell boxes which have handfuls of shells inside them.  The choice of adding 8 shotgun shell boxes is because they each hold 25 shells, thus 25 * 8 = 200.  I can adjust this by making the NanoAmmo dispense handfuls and do away with the pretense of shotgun shell boxes, but this would introduce a reverse issue when attempting to restock the shotgun shell boxes themselves.  I could address this in various ways but I'm not sure what you want from the feedback quite yet.

Cerrik replied:
I came up with a solution that is still scalable.  It simply deducts ammo from the shotgun ammo box to stock one shotgun shell box.

QualityVan reviewed (changes requested):
The whole thing with max_capacity and the vending machine directly adjusting record amounts should be in the vending_product datum and its attempt_restock instead. Put the max list on machinery/vending and pull from it during build_inventory, can default to -1 like how unlimited supplies do with amount.

Cerrik commented:
Changes addressed in recent commit.  There is a visual bug currently with stocking from ammo boxes, but that is out of scope for this PR and will be addressed with a different PR.

Cerrik commented:
Forgot to also mention that the single action army revolver ammo from the seasonal rotation is also excluded.  This is because the ammo for the SAA revolver is /obj/item/ammo_magazine/revolver, while every other revolver is a child, such as the normal R44 revolver ammo (/obj/item/ammo_magazine/revolver/standard_revolver)
This causes a unique issue with the new max_capacity var on the vending_product datum, which causes every other revolver ammo to not load correctly.  I could hack some fixes in, but I think the correct solution is for another PR where I move the SAA ammo to /obj/item/ammo_magazine/revolver/single_action/m44 or something like it.

yyzsong commented:
I can't wait to walk onto the tad after this PR is merged to see that the floors are still covered in 999 miscellaneous ammo boxes

Cerrik commented:
I adjusted the starting amount of every ammunition to be exactly half of their maximum capacity.

I also adjusted some select ammunition's max capacities to be higher, though these are the strictest max limits I would ever advise.

Specifically, shotgun shell boxes went from a max of 16 to a max of 32, (from 400 total shells to 800).  The reason I made this change is because every maximum capacity has been based on "How much ammo can a marine carry in both hands onto the tadpole?"  Initially, this was based on the shotgun ammo boxes from RO which hold 200 each, but I realized that custom ammo boxes can hold 400 shells alone.

There's some caveats to these maximums, such as how packets of .45-70 can have 32 packets in each ammo box so its max should actually be 64, but anyone who's ever used .45-70 knows you literally never run out of ammo anyway so I didn't see a reason to give them more.

Cerrik commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,191 @@
+		return
+
+	else if(istype(I, /obj/item/shotgunbox))
+		var/obj/item/shotgunbox/B = I
+		for(var/datum/vending_product/checked_record AS in product_records + hidden_records + coin_records)	// Loop through vendor records to find a match
```
Also borrowed the single letter vars from code this was built off of, only did it cause I thought it was some standard.  Will adjust!

Cerrik commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,191 @@
+/obj/machinery/vending/nanoammo/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/storage/box/visual/magazine))
+		var/obj/item/storage/box/visual/magazine/B = I
+		for(var/x = length(B.contents); x > 0; x--)
+			stock(B.contents[x], user, show_feedback = FALSE)
+		user?.balloon_alert(user, "\n\nYou dump the contents of the box into the NanoAmmo.");
```
Newbie understanding of DM code, will change

Cerrik commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,191 @@
+	)
+	mouse_over_pointer = MOUSE_HAND_POINTER
+
+/obj/machinery/vending/nanoammo/Initialize(mapload, ...)
+	. = ..()
+	switch(dir)
+		if(NORTH)
+			pixel_y = -14
+		if(SOUTH)
+			pixel_y = 26
+		if(EAST)
+			pixel_x = -19
+		if(WEST)
+			pixel_x = 21
```
Matches the same setup as the nanomed, so I assumed this was just the standard.  I'm not sure I know what mapping directional helpers means though, so I'll take this to discord

TiviPlus commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,181 @@
+		var/obj/item/storage/box/visual/magazine/ammo_box = I
+		for(var/mag in ammo_box.contents)
+			stock(mag, user, show_feedback = FALSE)
+		user?.balloon_alert(user, "\n\nYou dump the contents of the box into the NanoAmmo.");
```
Whats with the /n here

Cerrik replied:
Stupid mistake, left that in there from a previous test.  Will remove shortly.

TiviPlus commented on code/game/objects/machinery/vending/vending_types.dm:
```diff
@@ -408,6 +407,181 @@
+
+	else if(istype(I, /obj/item/shotgunbox))
+		var/obj/item/shotgunbox/big_shotgun_box = I
+		for(var/datum/vending_product/checked_record AS in product_records + hidden_records + coin_records)	// Loop through vendor records to find a match
+			var/obj/item/ammo_magazine/shotgun_shell_box = checked_record.product_path
+			if(big_shotgun_box.ammo_type == shotgun_shell_box.default_ammo)	// Found a match
```
```suggestion
		for(var/datum/vending_product/checked_record AS in product_records + hidden_records + coin_records)
			var/obj/item/ammo_magazine/shotgun_shell_box = checked_record.product_path
			if(big_shotgun_box.ammo_type == shotgun_shell_box.default_ammo)
```
Comments just describing what the line does, people can read (I hope)

Cerrik replied:
Those were the comments I pulled from the overridden function, figured I'd just keep the comments the same since they didn't hurt, but can remove them

Fawcksy commented:
very nice

Kuro020 commented:
Apparently this restores charges to TE batteries, is that intentional?
Its pretty stinky and shouldnt do that

Cerrik commented:
It does but that's because I asked about it very early in the PR for feedback and nobody answered so I just went with the default behavior of limited-item vendors.  The limited item checklist isn't very complete and lets a lot of things pass, such as partially charged TE stuff, so I can address this by fixing the limited checklist in another PR or just remove TE stuff from the nanoammo entirely given the field cell recharger already exists.

From earlier in the PR: "The current version allows TE cells, TE powerpacks, and TE powerbackpacks to be stored, up to a ridiculous amount of 60. I left them in because I considered them to be in that "basic ammo" category, but I have reservations given the TE field recharger also exists and this would partially negate it. I'll need some feedback on the TE category."

ViceSGT commented:
Add Nanoammo to new tad option please.
#17628
